### PR TITLE
WIP - start to explore data loader with subscriptions

### PR DESCRIPTION
**Pull request by bbakerman**
This is an VERY exploratory PR on using DataLoader with subscriptions

So far on some VERY early testing - it hangs - which means my suspicions the "level counting" the DataLoader Instrumentation does might be true - more work to confirm

### CODE REVIEW

**Comment by bbakerman** [comment]
Investigation - the reason the DataLoaderINtrumentation does not complete is that the context in `graphql.execution.instrumentation.dataloader.FieldLevelTrackingApproach#beginExecutionStrategy` is never called back on `onFieldValuesInfo` which calls `increaseHappenedOnFieldValueCalls` to markup state

In turn each of the event messages (with their field fetches) also do no call `onFieldValuesInfo` and hence not update the level tracking state and hence it never will dispatch because calls got started but never called back to say they completed

**Comment by bbakerman** [comment]
This PR gets Subscriptions to work with DataLoaders

one thing I am not happy about is this

```
        public void clearStateForSubscriptions() {
            expectedFetchCountPerLevel.clear();
            fetchCountPerLevel.clear();
            expectedStrategyCallsPerLevel.clear();
            happenedStrategyCallsPerLevel.clear();
            happenedOnFieldValueCallsPerLevel.clear();
            dispatchedLevels.clear();

            // level 1 for subscriptions is always completed - e.g. the subscription Publisher top level field
            // but every event after that starts from level 2
            expectedFetchCountPerLevel.set(1,1);
            fetchCountPerLevel.set(1,1);
            // with subscriptions, we don't re-enter the execution strategy
            // for levels 1 or 2 like we might on a normal run, so we set it up
            // pre-set so the call counts match
            expectedStrategyCallsPerLevel.set(1,1);
            expectedStrategyCallsPerLevel.set(2,1);
            happenedStrategyCallsPerLevel.set(1,1);
            happenedOnFieldValueCallsPerLevel.set(1,1);
            dispatchedLevels.add(1);
        }

````

The reason for this is as follows

The `DataLoaderDispatcherInstrumentation/ FieldLevelTrackingApproach` code tracks state of level across time.

So you have the first phase that is a normal graphql execution and a `Publisher` object is the result.

But then in subscriptions we keep pumping data down as that Publisher emits events.

but we start from a object on memory and mostly just "completeValue" on it and this runs enough Instrumentation calls back to work for the first event sent back.

BUT on the second event sent, the Dataloader state expects that you will enter a execution strategy again and that previous state matters        

eg you have dispatched level 0 and level 1 - which you have.

But it will have recorded that you have dispatched level 2,3,4...N for the first object send out.

Once that object is sent out, we need to reset the tracking levels.

However a quirk of the fact that we never re-enter the execution strategy for the "seocn Payload event from the publisher" means we have to reset the stats as

```
            expectedStrategyCallsPerLevel.set(1,1);
            expectedStrategyCallsPerLevel.set(2,1);
```

Which is weird and arbitrary BUT it reflects where we are up to.  Part of my does not like this - part of me says its a reality - this where were we are up to in terms of levels and strategy calls

**Comment by bbakerman** [comment]
OK I did a run through with the team on this PR and it became obviously that the state tracking that `FieldLevelTrackingApproach` does is INCOMPATIBLE with subscriptions.

Why?  Because of async event completion of each "published event" and the fact that there could be N subscribers to a subscription `Publisher` object

The field tracking works per request and in the query case, the levels monotonically increase and the field tracking state can only be updated once and in an on going manner.

With subscriptions however, for every event emitted on the Publisher, the graphql field sub selection happens AND hence the instrumentation methods run.

The returned values are async and hence can complete out of order

So the field tracking state for  object A might be at some level meanwhile the field tracking state for object  B might be at another level.  object B will be emitted after object A on the Publisher however it MAY complete to a value before object B

Therefore we need to have "field tracking state" per subscriber to the `Publisher` per object being emitted

The current field tracking state is per graphql request and this is a LOT different to what is described above.

In short we would need a complete new "Subscription ready field tracking mechanism".

In fact all instrumentation would need this "per subscriber / per object" state to be truly useful in Subscriptions.
